mono/mini/xdebug.c

```diff
@@ -270,15 +273,14 @@ static int xdebug_method_count;
  *
  *   Emit debugging info for METHOD into an assembly file which can be assembled
  * and loaded into gdb to provide debugging info for JITted code.
- * LOCKING: Acquires the loader lock.
  */
 void
 mono_save_xdebug_info (MonoCompile *cfg)
 {
 	MonoDebugMethodJitInfo *dmji;
 
 	if (use_gdb_interface) {
-		mono_loader_lock ();
+		mono_os_mutex_lock (&xdebug_lock);
```

Comment:
We are using the loader lock because its the only high level lock, complicated code such as this should not be called while holding a low level lock.

Reply:
The problem of using a high level lock is that all callers must be audited to be loader lock safe. And this turned out to be wrong.

Using the loader lock should always be frowned upon due to its auditing complexity.

Reply:
So the problem seems to be that we have a bunch of code in fill_runtime_generic_context () which is executed while holding the domain lock, this should not be done.